content/learning/operator/operator.md

```diff
@@ -65,10 +65,100 @@ Kubebuilder will generate the controller code only after generating CRDs, so eve
+
+```
+func someExamplePredicate() predicate.Predicate {
+    return predicate.Funcs{
+        CreateFunc: func(e event.UpdateEvent) bool {
+            // Here you can compare old version and new version of the watched object
+            return condition_to_evaluate
+        },
+        UpdateFunc: func(e event.UpdateEvent) bool {
+            // Here you can compare old version and new version of the watched object
+            return condition_to_evaluate
+        },
+        DeleteFunc: func(e event.DeleteEvent) bool {
+            // Here you can compare old version and new version of the watched object
+            return condition_to_evaluate
+        },
+    }
+}
+```
+
+A Predicate takes an event of a particular type (Create / Update / Delete) and returns true if the event should be processed by reconcile.
+
+3. Another problem that you can face relates to the creation and the update of resources in the cluster.
+Given that a controller must be _idempotent_ (superfluous reconcile has no effect, e.g. resources already present in the cluster must be updated and not recreated every reconcile), it is better to use [CreateOrUpdate](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.0/pkg/controller/controllerutil#CreateOrUpdate), a function that is intrinsically idempotent and it is part of the `controller-utils` package.
+In order to use it you have to create an initial resource composed only by ObjectMeta's data; then you have to pass a _mutate function_ as a parameter, which represents a function that works on the resource itself for example adding labels, modifying annotations and describing specifications.
+
+Here there is a CreateOrUpdate example for a VirtualMachine:
+```
+vm := v1.VirtualMachine{ObjectMeta: metav1.ObjectMeta{Name: "name", Namespace: "namespace"}}
+    _, err = ctrl.CreateOrUpdate(ctx, r.Client, &vm, func() error {
+        UpdateVirtualMachineSpec(&vm)
+        UpdateLabels(vm.Spec.Template.ObjectMeta.Labels)
+        return ctrl.SetControllerReference(resource, &vm, r.Scheme)
+	})
+```
+
+
+4. While updating a resources, sometimes applying an Update() on a resource could trigger errors like: "_operation cannot be fulfilled on resource, the object has been modified_".
+This means that the controller is using a different version of the resource from the one present in the cluster; this of course leads to a failure when updating the resource.
+To overcome this problem you can use the function Patch() instead of Update().
+In fact, Patch() does not make any check on the resource version; a possible problem consists in the possibility to lose some updates (since the changes in a resource are compared thanks to the version of the resource itself).
+However, the hypothetical loss of updates should not be a problem because of idempotence.
```

Comment:
This sentence is not clear to me. Why the idempotence prevents you from any losses?

Reply:
For sure the Patch function does not make any control to resource version. In theory a not aligned resource version can lead to an update loss. Unfortunately  we could not find any explanation of why it is possible to use that function without having troubles. What we thought is that since this problem happens only if you try to patch an older version of the cluster; the situation should be something like this: in the cluster you have the resource version 10 and you try to Patch it with a resource version 9. Thanks to the idempotence property of the operator it should not be a problem to update again the resource, without losing data at all.

Reply:
I believe this actually depends on the context, as sometimes it may be appropriate to use a patch operation and in some others not, as it can lead to loss of data. I feel this mainly boils down to whether you are supposed to be the only owner of the resource (hence you may only overwrite your own changes, and it may not matter thanks to idempotence), or parallel changes are possible (hence you risk to loose data). At the same time, also the error mentioned at the beginning of this section may, or may not, be a problem, as the best solution may just be to get a new version of the resource and retry (e.g. in a subsequent reconcile). I believe these aspects should be better investigated and this section to be adapted accordingly.

Reply:
It would be very interesting to analyze these aspects. Unfortunately we have already searched in the Golang documentation and some forums, but there is very few materials about this (actually, we found nothing). For sure your analysis is much more better then our, but on the basis of our experience we really do not know how to improve it